### PR TITLE
sig-instrumentation: Add charter document

### PR DESCRIPTION
**Pull request by brancz**
This PR is currently work in progress and meant for sig-instrumentation to iterate on its charter document.

@kubernetes/sig-instrumentation-misc

### CODE REVIEW

File: sig-instrumentation/charter.md [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+
+Covers best practices for cluster observability through metrics, logging, and events across all Kubernetes components and development of relevant components.
+
+### In scope
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
+
+### Out of scope
+
+Outline of things that could be confused as falling into this SIG but don't
```
Comment (coffeepac) [inline sig-instrumentation/charter.md]: Topics we don't cover:
- event (metric, log) processing
- prescribing what are error states

**Comment by philips** [comment]
Thank you for your patience as we work to your SIG charter merged.

Through the process of reviewing the [first 11 charters][charters] the Steering Committee (SC) found a lot of copy/paste language caused by our poorly designed initial charter template. To fix this there is a new [SIG charter template][template] which focuses on the main things we want to see in charters: scope and responsibilities.

Please consider updating to this charter template and putting focus into the scope and responsibilities. SC members will focus more on a deep review of scope more than anything else and using this template will help with that focus.

[charters]: https://docs.google.com/spreadsheets/d/1i_8oM1KlmXo90zIIgNTdOg-M6xesTggFLnTkgmscMlI/edit#gid=0
[template]: https://github.com/kubernetes/community/blob/master/committee-steering/governance/sig-charter-template.md

File: sig-instrumentation/charter.md (hunk 1 above)
Comment (philips) [reply sig-instrumentation/charter.md]: I think these are reasonable additions.

**Comment by brancz** [comment]
@coffeepac @philips I updated this PR with the new charter template and incorporated your comments. Let's go through this in today's sig-instrumentation meeting.

File: sig-instrumentation/charter.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+### Deviations from [sig-governance]
+
+- This SIG does not have Chairs.
+
+### Subproject Creation
+
+By SIG Technical Leads
+
+[sig-governance]: sig-governance.md
```
Comment (DirectXMan12) [inline sig-instrumentation/charter.md]: this doesn't end up at the right place

File: sig-instrumentation/charter.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
+
+[Link to SIG section in sigs.yaml][sigs.yaml].
+
+#### Code, Binaries and Services
+
+- Components required for any Kubernetes cluster in regards to observability.
+- Interfaces/API definitions required for any Kubernetes cluster in regards to observability.
+- Well established but optional components for Kubernetes clusters, if endorsed by member.
```
Comment (DirectXMan12) [inline sig-instrumentation/charter.md]: make this `members` (plural), and maybe add a note about having multiple people who can maintain.

File: sig-instrumentation/charter.md [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+- Guidance for instrumentation in order to ensure consistent and high quality instrumentation of core components.
+
+#### Cross-cutting and Externally Facing Processes
+
+n/a
```
Comment (DirectXMan12) [inline sig-instrumentation/charter.md]: two concerns here:

- being the "reviewers" for instrumentation (somewhat akin to api-reviewers)
- being involved in things like node monitoring, where we collaborate with another sig on the details of monitoring parts of kubernetes

File: sig-instrumentation/charter.md [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+This sig follows adheres to the Roles and Organization Management outlined in [sig-governance]
+and opts-in to updates and modifications to [sig-governance].
+
+### Additional responsibilities of Tech Leads
+
+- No additional responsibilities of Tech Leads
+
+### Deviations from [sig-governance]
+
+- This SIG does not have Chairs.
```
Comment (DirectXMan12) [inline sig-instrumentation/charter.md]: maybe say "the SIG does not have chairs separately from the technical leads"

File: sig-instrumentation/charter.md (hunk 2 above)
Comment (DirectXMan12) [reply sig-instrumentation/charter.md]: (it's a missing page for me)

Comment (philips) [reply sig-instrumentation/charter.md]: Sorry, oversight in the template https://github.com/kubernetes/community/pull/2423

File: sig-instrumentation/charter.md (hunk 4 above)
Comment (philips) [reply sig-instrumentation/charter.md]: agreed, I think people should come to this SIG for advice on what metrics to export and the best practices on the dimensions

File: sig-instrumentation/charter.md [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+- Guidance for instrumentation in order to ensure consistent and high quality instrumentation of core components.
+
+#### Cross-cutting and Externally Facing Processes
+
+n/a
+
+### Out of scope
+
+- Processing of signals. For example ingesting metrics, logs, events into external systems.
+- Predescribing what error states are.
```
Comment (philips) [inline sig-instrumentation/charter.md]: Prescribing? I don't know if predescribing works here. Maybe something like this: 

- Cataloging thresholds or other error states based on instrumentation e.g. alert configurations

File: sig-instrumentation/charter.md [hunk 7]
```diff
@@ -0,0 +1,52 @@
+# SIG Instrumentation Charter
+
+This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
+the Roles and Organization Management outlined in [sig-governance].
+
+## Scope
+
+Covers best practices for cluster observability through metrics, logging, and
```
Comment (philips) [inline sig-instrumentation/charter.md]: Scope should be what a SIG owns. So, perhaps:

"Owns best practices for cluster..."

File: sig-instrumentation/charter.md [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+events across all Kubernetes components and development of relevant components.
+
+### In scope
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
+
+[Link to SIG section in sigs.yaml][sigs.yaml].
+
+#### Code, Binaries and Services
```
Comment (philips) [inline sig-instrumentation/charter.md]: It would be worthwhile to link to the SIG's sub-projects https://github.com/kubernetes/community/tree/master/sig-instrumentation#subprojects

File: sig-instrumentation/charter.md [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+## Scope
+
+Covers best practices for cluster observability through metrics, logging, and
+events across all Kubernetes components and development of relevant components.
+
+### In scope
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
+
+[Link to SIG section in sigs.yaml][sigs.yaml].
```
Comment (DirectXMan12) [inline sig-instrumentation/charter.md]: should this have an actual link here?

Comment (philips) [inline sig-instrumentation/charter.md]: Probably better to link here since it is generated from sigs.yaml https://github.com/kubernetes/community/tree/master/sig-instrumentation#subprojects

**Comment by brancz** [comment]
@coffeepac @DirectXMan12 @philips All comments addressed. I think we're in a good shape. Let me know if there anything else. Removing `[WIP]` as I'd be happy with this state.

**Comment by philips** [comment]
PTAL @derekwaynecarr @spiffxp as the steering committee reviewers.

File: sig-instrumentation/charter.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
+
+#### Code, Binaries and Services
+
+[List of subprojects](https://github.com/kubernetes/community/tree/master/sig-instrumentation#subprojects)
+
+- Components required for any Kubernetes cluster in regards to observability.
+- Interfaces/API definitions required for any Kubernetes cluster in regards to observability.
+- Well established but optional components for Kubernetes clusters, if endorsed by members. Each component must have two or more members as maintainers.
```
Comment (philips) [inline sig-instrumentation/charter.md]: I don't understand this line. Can you explain further?

File: sig-instrumentation/charter.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+### In scope
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
+
+#### Code, Binaries and Services
+
+[List of subprojects](https://github.com/kubernetes/community/tree/master/sig-instrumentation#subprojects)
+
+- Components required for any Kubernetes cluster in regards to observability.
+- Interfaces/API definitions required for any Kubernetes cluster in regards to observability.
```
Comment (philips) [inline sig-instrumentation/charter.md]: Can you link to those APIs and Interfaces that this SIG owns?

File: sig-instrumentation/charter.md [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+### In scope
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
+
+#### Code, Binaries and Services
+
+[List of subprojects](https://github.com/kubernetes/community/tree/master/sig-instrumentation#subprojects)
+
+- Components required for any Kubernetes cluster in regards to observability.
```
Comment (philips) [inline sig-instrumentation/charter.md]: What are those components? Are they the subprojects or something else? Can you link to them?

**Comment by brancz** [comment]
@philips I linked out to projects and prospects as examples for each category. Let me know if you like this format or if you have other suggestions.

File: sig-instrumentation/charter.md [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+This sig follows adheres to the Roles and Organization Management outlined in [sig-governance]
+and opts-in to updates and modifications to [sig-governance].
+
+### Additional responsibilities of Tech Leads
+
+- No additional responsibilities of Tech Leads
+
+### Deviations from [sig-governance]
+
+The SIG does not have Chairs separately from the Technical Leads.
```
Comment (spiffxp) [inline sig-instrumentation/charter.md]: I would rephrase this as "Tech Leads must also fulfill all of the responsibilities of the Chair role as outlined in [sig-governance]"

You may also want to consider whether Tech Leads also doing Chair duties just happens to be the state of today, or something you wish to require of all future Tech Leads

File: sig-instrumentation/charter.md [hunk 14]
```diff
@@ -0,0 +1,55 @@
+# SIG Instrumentation Charter
+
+This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
+the Roles and Organization Management outlined in [sig-governance].
+
+## Scope
+
+Owns best practices for cluster observability through metrics, logging, and
+events across all Kubernetes components and development of relevant components.
```
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: please just strike the term "events" from the charter.  it has not been historically the role of this sig and causes confusion.

File: sig-instrumentation/charter.md [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+
+### In scope
+
+Coordinating metric requirements of different SIGs for other components through finding common APIs.
```
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: historically to my knowledge, it has just been metrics and not logs or events.

File: sig-instrumentation/charter.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: to clarify more on events:

- the event API itself is owned by sig-arch imho given its cross-cutting concern; it predates sigs.
- the client side compression algo was done by me before formation of sig, but its part of client-go so api-machinery owns it now.
- the event v2 api proposal is grouped under sig-instrumentation at this time, but in my opinion its addressing the api from a scalability perspective.

see: https://github.com/kubernetes/community/blob/f367271772f6616088473a6d99a0f084266d4047/contributors/design-proposals/instrumentation/events-redesign.md

all this to say, events and their best practices are not owned by a single sig imho.  its by definition cross-cutting.

File: sig-instrumentation/charter.md [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+
+#### Code, Binaries and Services
+
+[List of subprojects](https://github.com/kubernetes/community/tree/master/sig-instrumentation#subprojects)
```
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: @spiffxp i think addons a conflated with a concept of a distribution, and i am not sure this sig should be the owning body for all of those.  defer to @brancz for his input, but i think its a grey area.

File: sig-instrumentation/charter.md [hunk 17]
```diff
@@ -0,0 +1,55 @@
...
+
+- Components required for any Kubernetes cluster in regards to observability. ([kubernetes-incubator/metrics-server](https://github.com/kubernetes-incubator/metrics-server), [kubenernetes/heapster](https://github.com/kubernetes/heapster))
```
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: i think this is clear enough, but we could say "core metrics pipeline" that used by core components (scheduler, kubectl, etc.)

File: sig-instrumentation/charter.md [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+- Interfaces/API definitions required for any Kubernetes cluster in regards to observability. ([kubernetes/metrics](https://github.com/kubernetes/metrics), [kubernetes-incubator/custom-metrics-apiserver](https://github.com/kubernetes-incubator/custom-metrics-apiserver))
```
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: this is really api extension points for supporting metric sources consumed by core k8s components (i.e. hpa consuming custom metric sources via api) where the metric source itself is not a k8s project (i.e. prometheus, etc.)

File: sig-instrumentation/charter.md [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+  - Reviewing any instrumentation related changes and additions.
+  - Guidance on what should be instrumented as well as dimensions of the same.
+  - Creating, adding and maintaining the Kubernetes instrumentation guidelines.
+  - Coordinate cross SIG instrumentation efforts.
+
+### Out of scope
+
+- Processing of signals. For example ingesting metrics, logs, events into external systems.
+- Dictating what states must result in an alert. Suggestions or opt-in alerts may be in scope.
+
```
[the remark this answers is not in the document]
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: i would say alert management is outside the scope entirely?

File: sig-instrumentation/charter.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (coffeepac) [reply sig-instrumentation/charter.md]: logging has been a concern owned by this SIG since at least January 2017 and there are e2e tests owned by this sig that predate that.

File: sig-instrumentation/charter.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (coffeepac) [reply sig-instrumentation/charter.md]: we've been discussing have a base set of 'suggested alerts' around core kubernetes functionality.  things like 99% API response time, pod startup time and then make it very clear that the actual thresholds need to be tuned by the user but provide some sane defaults.  would that be too prescriptive?

File: sig-instrumentation/charter.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (brancz) [reply sig-instrumentation/charter.md]: Correct, but it's wildly disputed what "correct" APIs for logging could and should look like, this goes from logging libraries (glog) to the actual format outputted by these. The only thing about logging that sig-instrumentation somewhat owns is the interface of log file directory structure. It neither owns the actual act of writing the files (typically done by the container runtime) nor does it dictate a format or how logs are to be processed afterwards. I believe we do own and maintain optional addons for this though.

File: sig-instrumentation/charter.md [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+- Interfaces/API definitions required for any Kubernetes cluster in regards to observability. ([kubernetes/metrics](https://github.com/kubernetes/metrics), [kubernetes-incubator/custom-metrics-apiserver](https://github.com/kubernetes-incubator/custom-metrics-apiserver))
+- Well established but optional components or adapters for Kubernetes clusters, if endorsed by members. Each component must have two or more members as maintainers. ([kubernetes/kube-state-metrics](https://github.com/kubernetes/kube-state-metrics), not yet officially owned by SIG Instrumentation, but an example prospect for this category: [DirectXMan12/k8s-prometheus-adapter](https://github.com/DirectXMan12/k8s-prometheus-adapter))
+
+#### Cross-cutting and Externally Facing Processes
+
+- Guidance for instrumentation in order to ensure consistent and high quality instrumentation of core Kubernetes components. This includes:
+  - Reviewing any instrumentation related changes and additions.
+  - Guidance on what should be instrumented as well as dimensions of the same.
+  - Creating, adding and maintaining the Kubernetes instrumentation guidelines.
```
[the remark this answers is not in the document]
Comment (brancz) [reply sig-instrumentation/charter.md]: It does :) https://github.com/kubernetes/community/blob/master/contributors/devel/instrumentation.md

File: sig-instrumentation/charter.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (brancz) [reply sig-instrumentation/charter.md]: None of the above fall into the "metrics pipeline" category per se, a subset of them could be selected to build one. fluentd-gcp and metadata-agent are or at least should be owned by sig-gcp, both depend on external third party services. Regarding the others:

* cluster-monitoring: is owned by this sig, but deprecated due to heapster deprecation
* fluentd-elasticsearch: we never discussed this formally, but I believe this is maintained by sig-instrumentation members, and I'm ok with that.
* prometheus: this is owned and maintained by us, however as its readme says: "This add-on is an experimental configuration of k8s monitoring using Prometheus used for e2e tests."

What do you want to do with this information? Should we add the addons we own in the sigs.yaml? Generally all of these addons fit into the "Well established but optional components or adapters for Kubernetes clusters" category.

File: sig-instrumentation/charter.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (brancz) [reply sig-instrumentation/charter.md]: Yes, what @coffeepac said is correct. We have been discussing that suggestions for alerts could be something we can own, as that it already falls into most members daily work, and this could allow to centralize suggested alert sets more and better.

**Comment by brancz** [comment]
Added further updates according to the comments. Please take a look.

**Comment by philips** [comment]
@spiffxp and @derekwaynecarr if you think the SIG Instrumentation charter now looks good can you please approve?

**Comment by spiffxp** [comment]
looking...

File: sig-instrumentation/charter.md [hunk 21]
```diff
@@ -0,0 +1,69 @@
+# SIG Instrumentation Charter
+
+This charter adheres to the conventions described in the [Kubernetes Charter README] and uses
+the Roles and Organization Management outlined in [sig-governance].
+
+## Scope
+
+Owns best practices for cluster observability through metrics and logging
+across all Kubernetes components and development of components required for all
```
Comment (spiffxp) [inline sig-instrumentation/charter.md]: I'm still not sure "Kubernetes components" is a well understood, well defined term, but I will assume you are OK deferring the definition of this term to SIG Architecture and accepting whatever scope that implies if the definition is refined.

File: sig-instrumentation/charter.md [hunk 22]
```diff
@@ -0,0 +1,69 @@
...
+  - Guidance on what should be instrumented as well as dimensions of the same. (see [Instrumenting Kubernetes Guide][instrumenting-kubernetes])
+  - Creating, adding and maintaining the Kubernetes instrumentation guidelines.
+  - Coordinate cross SIG-Instrumentation efforts.
+
+### Out of scope
+
+- Processing of signals. For example ingesting metrics, logs, events into external systems.
+- Dictating what states must result in an alert. Suggestions or opt-in alerts may be in scope.
+- Cloud provider specific addons are out of scope and should be taken care of by the respective SIG.
+
```
Comment (spiffxp) [inline sig-instrumentation/charter.md]: I feel like this should contain some of @brancz's response
> It neither owns the actual act of writing the files (typically done by the container runtime) nor does it dictate a format or how logs are to be processed afterwards.

eg: 
- The act of writing log files, their format, or how they are to be processed afterwards

File: sig-instrumentation/charter.md [hunk 23]
```diff
@@ -0,0 +1,69 @@
...
+The act of instrumenting components not owned by SIG-Instrumentation is out of
+scope, however SIG-Instrumentation is there to advise any contributors with
+instrumentation decisions.
+
+As well as giving advise in regards to instrumentation, SIG-Instrumentation
+coordinates metric requirements of different SIGs for other components through
+finding common APIs (such as the resource/core, custom and external metrics
+APIs).
+
```
Comment (spiffxp) [inline sig-instrumentation/charter.md]: This seems relevant as well

>The only thing about logging that sig-instrumentation somewhat owns is the interface of log file directory structure.

File: sig-instrumentation/charter.md [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+#### Code, Binaries and Services
+
+[List of subprojects](https://github.com/kubernetes/community/tree/master/sig-instrumentation#subprojects)
```
Comment (spiffxp) [inline sig-instrumentation/charter.md]: >Should we add the addons we own in the sigs.yaml?

Yes, I'll open a follow on PR based on your input, it needn't block this

File: sig-instrumentation/charter.md [hunk 25]
```diff
@@ -0,0 +1,69 @@
...
+### In scope
+
+SIG-Instrumentation revolves around the process of instrumenting and exposing
+observability signals.
+
+The act of instrumenting components not owned by SIG-Instrumentation is out of
+scope, however SIG-Instrumentation is there to advise any contributors with
+instrumentation decisions.
+
+As well as giving advise in regards to instrumentation, SIG-Instrumentation
```
Comment (spiffxp) [inline sig-instrumentation/charter.md]: nit: s/advise/advice

**Comment by spiffxp** [comment]
If I see no objections by end of day I'm going to re-add the LGTM per the above

**Comment by philips** [comment]
sounds good to me @spiffxp

File: sig-instrumentation/charter.md (hunk 21 above)
Comment (derekwaynecarr) [reply sig-instrumentation/charter.md]: i am fine w/ that caveat as described by @spiffxp

File: sig-instrumentation/charter.md [hunk 26]
```diff
@@ -0,0 +1,73 @@
...
+The act of instrumenting components not owned by SIG-Instrumentation is out of
+scope, however SIG-Instrumentation is there to advise any contributors with
+instrumentation decisions.
+
+As well as giving advice in regards to instrumentation, SIG-Instrumentation
+coordinates metric requirements of different SIGs for other components through
+finding common APIs (such as the resource/core, custom and external metrics
+APIs).
+
+SIG Instrumentation owns the interface of the log file and directory structure
```
Comment (derekwaynecarr) [inline sig-instrumentation/charter.md]: we did not discuss this in the sig-node charter. 

i think this should have some further refinement.

at best I think this is with sig-node, and not exclusively instrumentation.  the sig-node charter says "Issues related to node, pod, container monitoring (with [sig-instrumentation](../sig-instrumentation))" and probably should have also had a line for logging w/ sig-instrumentation.  I think this is shared responsibility across the two sigs.

Comment (brancz) [reply sig-instrumentation/charter.md]: Makes sense. Will move that to shared responsibilities.

Comment (brancz) [reply sig-instrumentation/charter.md]: I deleted this line and added: "The interface of log files and their directory structure written out by container runtimes to be processed by other systems further, is shared responsibility between [SIG Node](sig-node) and SIG Instrumentation." To the cross-cutting section.
